docker/build-tools/Dockerfile

```diff
@@ -57,45 +62,38 @@ RUN go get github.com/jteeuwen/go-bindata/go-bindata
 RUN go get istio.io/tools/protoc-gen-docs
 RUN go get istio.io/tools/cmd/annotations_prep
 
-# Save some bytes in the container
-RUN upx --lzma /go/bin/*
+# Put the stuff we need in its final output location
+RUN mkdir -p ${OUTDIR}/go/bin
+RUN mkdir -p ${OUTDIR}/usr/include
 
-# Cleanup
-WORKDIR /
-RUN apt-get purge -y \
-    autoconf \
-    libc-ares-dev \
-    libssl-dev \
-    libtool \
-    unzip \
-    upx \
-    zlibc \
-    zlib1g-dev \
-    && apt-get autoremove -y \
-    && apt-get clean \
-    && rm -rf /var/lib/apt/lists/* \
-    && rm -fr /home /tmp/* \
-    && go clean -cache \
-    && go clean -modcache
+RUN cp -aR /tmp/bin/protoc ${OUTDIR}/go/bin
+RUN cp -aR /tmp/include/* ${OUTDIR}/usr/include
+RUN cp -aR /go/bin/* ${OUTDIR}/go/bin
+
+# Use inplace decompression on the go toolchain binaries
+RUN upx --lzma ${OUTDIR}/go/bin/*
+
+# Create a golang toolchain context
+FROM golang:1.12 as golang_context
+
+# Build a base os context
+FROM ubuntu:xenial as base_os_context
+
+RUN apt-get update
+RUN apt-get install -y make ca-certificates git
+RUN apt-get autoremove -y
+RUN apt-get clean
+RUN rm -rf /var/lib/apt/lists/*
 
 # Setup to run the Go compiler with host-side build cache, test cache, and module cache
+FROM scratch
+
+ENV GOROOT=/go
```

Comment:
this is wrong and should be deleted.  This explains why I needed a bindmount to $HOME/go/src, rather than `/go` - as /go/mod/pkg is bindmounted to $HOME/gho/mod/pkg when $GOROOT is set to `/go`.